runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1175,3 +1176,163 @@ J9::ValuePropagation::getParmValues()
    }
+
+static bool isTakenSideOfAVirtualGuard(TR::Block* block)
+   {
+   TR::TreeTop* lastTT = block->getLastRealTreeTop();
+   for (TR::TreeTop* tt = block->getEntry(); tt != lastTT; tt = tt->getNextTreeTop())
+      {
+      if (tt->getNode()->getNumChildren() > 0
+         && tt->getNode()->getFirstChild()->isTheVirtualCallNodeForAGuardedInlinedCall())
+         return true;
+      }
+   return false;
+   }
+
+static bool skipFinalFieldFoldingInBlock(TR::Block* block)
+   {
+   if (block->isCold() ||
+       TR_FearPointAnalysis::shouldSkipBlock(block) ||
+       isTakenSideOfAVirtualGuard(block))
+      return true;
+
+   return false;
+   }
+
+static bool isStaticFinalFieldWorthFolding(TR::Compilation* comp, TR_OpaqueClassBlock* declaringClass, char* fieldSignature, int32_t fieldSigLength)
+   {
+   if (comp->getMethodSymbol()->hasMethodHandleInvokes()
+       && !TR::Compiler->cls.classHasIllegalStaticFinalFieldModification(declaringClass))
+      {
+      if (fieldSigLength == 28 && !strncmp(fieldSignature, "Ljava/lang/invoke/VarHandle;", 28))
+         return true;
+      }
+
+   return false;
+   }
+
+
+// Do not add fear point in a frame that doesn't support OSR
+static bool safeToAddFearPointIn(TR::Compilation* comp, int32_t callerIndex)
+   {
+   TR::ResolvedMethodSymbol *method = callerIndex == -1 ?
+      comp->getJittedMethodSymbol() : comp->getInlinedResolvedMethodSymbol(callerIndex);
+
+   return !method->cannotAttemptOSRDuring(callerIndex, comp, false);
+   }
+
+bool J9::ValuePropagation::transformDirectLoad(TR::Node* node)
+   {
+   // Allow OMR to fold reliable static final field first
+   if (OMR::ValuePropagation::transformDirectLoad(node))
+      return true;
+
+   if (node->isLoadOfStaticFinalField() &&
+       tryFoldStaticFinalFieldAt(_curTree, node))
+      {
+      return true;
+      }
+
+   return false;
+   }
+
+/** \brief
+ *     Try to fold static final field
+ *
+ *  \param tree
+ *     The tree with the load of static final field.
+ *
+ *  \param fieldNode
+ *     The node which is a load of a static final field.
+ */
+bool J9::ValuePropagation::tryFoldStaticFinalFieldAt(TR::TreeTop* tree, TR::Node* fieldNode)
+   {
+   TR_ASSERT(fieldNode->isLoadOfStaticFinalField(), "Node n%dn %p has to be a load of a static final field", fieldNode->getGlobalIndex(), fieldNode);
+
+   if (comp()->getOption(TR_DisableGuardedStaticFinalFieldFolding))
+      {
+      return false;
+      }
+
+   if (!comp()->supportsInduceOSR() ||
+       !comp()->isOSRTransitionTarget(TR::postExecutionOSR) ||
+       !comp()->getOSRMode() == TR::voluntaryOSR)
+      {
+      return false;
+      }
+
+   if (fieldNode->getByteCodeInfo().doNotProfile() ||
+       skipFinalFieldFoldingInBlock(tree->getEnclosingBlock()) ||
+       !safeToAddFearPointIn(comp(), fieldNode->getByteCodeInfo().getCallerIndex()) ||
+       TR::TransformUtil::canFoldStaticFinalField(comp(), fieldNode) != TR_maybe)
```

Comment:
I think this would read better with the ||s a the start of the lines to make the condition relationships more obvious since this is quite complicated.

Reply:
Fixed.